pkg/replicator/scheme.go

```diff
@@ -0,0 +1,387 @@
+		return fmt.Errorf("read origin meta file: %w", err)
+	}
+
+	if targetMetaFile != nil && !rs.toBkt.IsObjNotFoundErr(err) {
+		targetMetaFileContent, err := ioutil.ReadAll(targetMetaFile)
+		if err != nil {
+			return fmt.Errorf("read target meta file: %w", err)
+		}
+
+		if bytes.Equal(originMetaFileContent, targetMetaFileContent) {
+			// If the origin meta file content and target meta file content is
+			// equal, we know we have already successfully replicated
+			// previously.
+			level.Debug(rs.logger).Log("msg", "skipping block as already replicated", "block_uuid", id.String())
+			rs.metrics.blocksAlreadyReplicated.Inc()
+
+			return nil
+		}
+	}
+
+	if err := rs.fromBkt.Iter(ctx, chunksDir, func(objectName string) error {
+		err := rs.ensureObjectReplicated(ctx, objectName)
+		if err != nil {
+			return fmt.Errorf("replicate object %v: %w", objectName, err)
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if err := rs.ensureObjectReplicated(ctx, indexFile); err != nil {
+		return fmt.Errorf("replicate index file: %w", err)
+	}
+
+	level.Debug(rs.logger).Log("msg", "replicating meta file", "object", metaFile)
+
+	if err := rs.toBkt.Upload(ctx, metaFile, bytes.NewReader(originMetaFileContent)); err != nil {
+		return fmt.Errorf("upload meta file: %w", err)
+	}
+
+	rs.metrics.blocksReplicated.Inc()
+
+	return nil
+}
+
+// ensureBlockIsReplicated ensures that an object present in the origin bucket
+// is present in the target bucket.
+func (rs *replicationScheme) ensureObjectReplicated(ctx context.Context, objectName string) error {
+	level.Debug(rs.logger).Log("msg", "ensuring object is replicated", "object", objectName)
+
+	exists, err := rs.toBkt.Exists(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("check if %v exists in target bucket: %w", objectName, err)
+	}
+
+	// skip if already exists.
+	if exists {
+		level.Debug(rs.logger).Log("msg", "skipping object as already replicated", "object", objectName)
+		return nil
+	}
+
+	level.Debug(rs.logger).Log("msg", "object not present in target bucket, replicating", "object", objectName)
+
+	r, err := rs.fromBkt.Get(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("get %v from origin bucket: %w", objectName, err)
+	}
+
+	defer r.Close()
+
+	if err = rs.toBkt.Upload(ctx, objectName, r); err != nil {
+		return fmt.Errorf("upload %v to target bucket: %w", objectName, err)
+	}
+
+	level.Info(rs.logger).Log("msg", "object replicated", "object", objectName)
+	rs.metrics.objectsReplicated.Inc()
+
+	return nil
+}
+
+// loadMeta loads the meta.json from the origin bucket and returns the meta
+// struct as well as if failed, whether the failure was due to the meta.json
+// not being present or partial. The distinction is important, as if missing or
+// partial, this is just a temporary failure, as the block is still being
+// uploaded to the origin bucket.
+func loadMeta(ctx context.Context, bucket objstore.BucketReader, id ulid.ULID) (*metadata.Meta, bool, error) {
```

Comment:
I feel like you could use our `MetaFetcher` here and for the 2nd return argument you could check if the returned error is `block.ErrorSyncMetaNotFound`.